### ISSUE TITLE
"roundabout" announcements improvements

### ISSUE THREAD

**Issue by wavexx** [osmandapp/Osmand#1603]
By living in EU (France/Italy), I find the "prepare to enter" roundabout announcements totally worthless. There's nothing special about roundabouts compared to a crossing, and in cities you might have an average higher than one per square kilometer, which makes the TTS so noisy I had to turn off announcements.

The exit announcement is good, but there's no need to a "prepare to enter a roundabout" warning. Roundabouts are even more prominently marked than a crossing, with slow-down panels as early as 1km in certain cases.

Not only that, but I find the anticipation of the announcement to be completely off. I've had several times "in 4 km, prepare to enter a roundabout".

I've worked for a custom a street-nav before, and one of the best solutions we found was to give announcements a time-based anticipation (based on current distance/speed) plus a hard-limit distance (10/50/100/...m depending on the announcemen + GPS acc radius), just to ensure the announcement is not skipped.

I care about regular turning announcements only 100 meters or 15 seconds before the actual turn, whichever is first. There's no need to announce a turn ~50km away just because it's the next turn. I gladly prefer the TTS to stay shut if all I need to do is go straight.

[Another thing we did, to improve the TTS in eu countries, was to detect the angle of the exit turn in roundabouts. What happens frequently is the transformation of a crossing in a roundabout: in this case, you have a large number of 4-exit roundabouts where the exit points are almost at straight angles. When the roundabout has 4 exit, and the 2nd exit is chosen, and the angle from the current segment/exit is <5deg, we replace the accouncment with "at the next roundabout, continue straight with the 2nd exit"]

**Comment by sonora** [comment]
Valid point, let's focus on this in Issue 1411.

**Comment by sonora** [comment]
I leave this open for the roundabout angle detection, while the lead time / lead distance fine tuning should be done in one go in #1411.

**Comment by vshcherb** [comment]
Well we don't have currently possibility to properly detect turn angle, we refused that idea on display screen and now we are just going to show exit number. Some roundabouts are very tricky to calculate proper turn-angle.

I guess wording will be addressed in 1411, so I will close that issue.

**Comment by wavexx** [comment]
We referred to "the turn angle" as the angle between the entry and the exit road, disregarding the geometry of the roundabout itself.

Since roundabouts are often not "regular", to calculate the angle of the roads on the plane, we take the angular vector of the road from the node touching the roundabout to a point on the exiting road at 25 meters, or at the the first intersecting road (whichever comes first).

![img_20150903_154528828](https://cloud.githubusercontent.com/assets/1017726/9659934/5df47fc4-5253-11e5-950c-b7543336b70f.jpg)

To illustrate with this crappy sketch, imagine coming from the road at the bottom. We calculate the angle of the road (alpha), from the node touching the roundabout (1) to 25m outwards (x). We repeat the step on each exiting road. The angle of the exits are determined by alpha-beta, alpha-gamma, etc. The geometry of the roundabout becomes irrelevant, and this will work also for non-circular roundabouts.

As an exception, we also took into account the size of the roundabout (the length of all the segments being part of the roundabout). If above 1km, we ignore the angle detection (this happens in very *big*, non-regular roundabouts), since the driver will have quite a while before exiting, and by that time the angle of exit becomes irrelevant.
